File: src/gestorAplicacion/personas/Paciente.java
package gestorAplicacion.personas;
import gestorAplicacion.Cita;
import gestorAplicacion.Consulta;
import gestorAplicacion.Examen;
import gestorAplicacion.estadoEntrega;
import gestorAplicacion.tipoCita;
import gestorAplicacion.tipoExamen;
import gestorAplicacion.registrosMedicos.Diagnostico;
import gestorAplicacion.registrosMedicos.HistoriaClinica;
import gestorAplicacion.Consultorio;
import gestorAplicacion.Entrega;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Map;
import java.util.TreeMap;
import java.time.format.DateTimeFormatter;
import java.util.Scanner;



public class Paciente extends Persona{
	private static final long serialVersionUID = 1L;
	private String eps;
	private ArrayList<String> sintomas;
	private HistoriaClinica historiaClinica;
	private Map<LocalDateTime, Consulta> consultas = new TreeMap<LocalDateTime, Consulta>();
	private Map<LocalDateTime, Entrega> entregas = new TreeMap<LocalDateTime, Entrega>();
	public boolean pagado;
    public static int pacientes;

	//Constructores
	public Paciente(String nombre, String apellido, tipoDocumento documento, long numeroDocumento, int edad, String genero, long telefono, String correoElectronico, String direccion, String eps, 
			ArrayList<String> sintomas, HistoriaClinica historiaClinica, boolean pagado) {
		super(nombre, apellido, documento, numeroDocumento, edad, genero, telefono, correoElectronico, direccion);
		setEps(eps);
		setSintomas(sintomas);
		setHistoriaClinica(historiaClinica);
		this.pagado = pagado;
                pacientes++;
		
	}
	
	public Paciente() {
		this(null,null, null, 0, 0, null, 0, null, null,null, new ArrayList<String>(), null, false);
	}

	//Getters y setters
	public String getEps() {
		return eps;
	}
	public void setEps(String eps) {
		this.eps = eps;
	}
	
	public ArrayList<String> getSintomas() {
		return sintomas;
	}
	public void setSintomas(ArrayList<String> sintomas) {
		this.sintomas = sintomas;
	}
	
	public HistoriaClinica getHistoriaClinica() {
		return historiaClinica;
	}
	public void setHistoriaClinica(HistoriaClinica historiaClinica) {
		this.historiaClinica = historiaClinica;
	}
	public Map<LocalDateTime, Consulta> getConsultas() {
		return consultas;
	}

	public void setConsultas(Map<LocalDateTime, Consulta> consultas) {
		this.consultas = consultas;
	}

	public boolean isPagado() {
		return pagado;
	}

	public void setPagado(boolean pagado) {
		this.pagado = pagado;
	}
		
	public Map<LocalDateTime, Entrega> getEntregas() {
		return entregas;
	}

	public void setEntregas(Map<LocalDateTime, Entrega> entregas) {
		this.entregas = entregas;
	}

	//Metodos
	public String pedirCita(LocalDateTime fecha, tipoCita tipoCita, String motivo) {
		ArrayList<Medico> medicos = Medico.medicos;
		Medico medico = null;
		for (int i = 0; i<medicos.size(); i++){
			//Se recorre la lista de los medicos creados y se va verificando si la especialidad del medico i concuerda con el tipo de cita que
			//se requiere y tambien se verifica si ese medico esta disponible.
			if ( medicos.get(i).especialidad == tipoCita && Administrador.verificarDisponibilidadMedico(fecha, medicos.get(i)) == true) {
				medico = medicos.get(i);
				break;
			}
		}
		ArrayList<Consultorio> consultorios = Consultorio.consultorios;
		Consultorio consultorio = null;
		for (int i = 0; i< consultorios.size(); i++) {
			//Se recorre la lista de los consultorios creados y se verifica si el consultorio esta disponible en la fecha que se requiere la cita
			if(Administrador.verificarDisponibilidadConsultorio(fecha, consultorios.get(i)) == true){
				consultorio = consultorios.get(i);
				break;
			}
		}
		if (medico != null && consultorio != null) {
			DateTimeFormatter formato = DateTimeFormatter.ofPattern("dd-MM-yyyy HH:mm");  
			Administrador.asignarCita(this, medico, consultorio, fecha, motivo, tipoCita);
			return "Su cita ha sido agendada exitosamente el: " + fecha.format(formato) + " con el médico: "+ medico.getNombre()+" "+ medico.getApellido()+ " en el consultorio "+ consultorio.getId()+"\n";
		}
		else {
			return "Hubo un error agendando su cita, vuelvalo a intentar. \n";
		}
	}
	
	public boolean reagendarCita(LocalDateTime fecha) {
		Cita cita = null;
		ArrayList<Cita> citas = Cita.listaCitas;
		//Se busca la cita que se quiere reagendar
		for (int i = 0; i<citas.size(); i++) {
			if (citas.get(i).getPaciente() == this) {
				cita = citas.get(i);
				break;
			}
		}
		//Se verifica que tanto el medico como el consultorio esten disponible en la nueva fecha
		if(Administrador.verificarDisponibilidadMedico(fecha,cita.getMedico()) == true && Administrador.verificarDisponibilidadConsultorio(fecha, cita.getConsultorio())==true){
			cita.setFecha(fecha);
			return true;
		}
		else {
			return false;
		}
	}
	
	//En caso de que el medico no este disponible
	public boolean reagendarCitaMedico(LocalDateTime fecha) {
		Cita cita = null;
		ArrayList<Cita> citas = Cita.listaCitas;
		for (int i = 0; i<citas.size(); i++) {
			if (citas.get(i).getPaciente() == this) {
				cita = citas.get(i);
				break;
			}
		}
		//Se busca un medico con disponibilidad
		ArrayList<Medico> medicos = Medico.medicos;
		Medico medico = null;
		for(int i = 0; i<medicos.size(); i++) {
			if(medicos.get(i).especialidad ==  cita.getTiposCitas() && Administrador.verificarDisponibilidadMedico(fecha, medicos.get(i)) == true){
				medico = medicos.get(i);
				cita.setMedico(medico);
				cita.setFecha(fecha);
				break;
			}
			else if(Administrador.verificarDisponibilidadMedico(fecha, medicos.get(i)) == false && i== medicos.size()){
				return false;
			}
		}
		//Se verifica si el consultorio esta disponible para la nueva fecha
		if (Administrador.verificarDisponibilidadConsultorio(fecha, cita.getConsultorio())==true && medico != null) {
			return true;
		}
		//Si el consultorio no esta disponible, se busca un nuevo consultorio
		else {
			ArrayList<Consultorio> consultorios = Consultorio.consultorios;
			Consultorio consultorio = null;
			for (int i = 0; i< consultorios.size(); i++) {
				//Se recorre la lista de los consultorios creados y se verifica si el consultorio esta disponible en la fecha que se requiere la cita
				if(Administrador.verificarDisponibilidadConsultorio(fecha, consultorios.get(i)) == true){
					consultorio = consultorios.get(i);
					cita.setConsultorio(consultorio);
					break;
				}
			}
			return true;
		}
	}

	
	public String solicitarExamen(tipoExamen tipo, LocalDateTime fecha) {
		Examen examen = new Examen((int)(Math.random()*10000+1), this, tipo, false);
		Administrador.autorizarExamen(examen);
		// Filtrar médicos disponibles para la fecha
		// Filtrar consultorios disponibles para la fecha
		return Administrador.asignarExamen(examen, this, Medico.medicos, Consultorio.consultorios, fecha);
	}
	
	
//	public String cancelarCita(LocalDateTime fecha, tipoCita tipoCita) {
//		Cita cita = null;
//		ArrayList<Cita> citas = Cita.listaCitas;
//		//Se busca la cita que se quiere reagendar
//		for (int i = 0; i<citas.size(); i++) {
//			if (citas.get(i).getPaciente() == this && citas.get(i).getFecha() == fecha && citas.get(i).getTiposCitas() == tipoCita) {
//				cita = citas.get(i);
//				break;
//			}
//		}
//		cita.cia
//	}
	
	public HistoriaClinica consultarHistorioaClinica() {
		return getHistoriaClinica();
	}
//	public void solicitarExamen(tipoExamen tipo, LocalDateTime fecha) {
//		Examen examen = new Examen((int)(Math.random()*10000+1), this, tipo, false);
//		Administrador.autorizarExamen(examen);
//		// Filtrar médicos disponibles para la fecha
//		// Filtrar consultorios disponibles para la fecha
//		Administrador.asignarExamen(examen, this, medicos, consultorios, fecha);
//
//	}

	public static String solicitarEnvioMedicamento (String nombre, String domicilio){
		ArrayList<Diagnostico> lis = Medico.listado;
		int cont = 1;
		for (Diagnostico medi : lis) {
			if (medi.getPersona().getNombre().equals(nombre)){
				String option;
				if (medi.getPersona().pagado == true){
					Entrega.crearEntrega(cont, medi.getPersona().getNombre(),domicilio,medi.getMedicamiento(), estadoEntrega.En_camino);
					cont++;
					return "el proceso ha sido exitoso";
				}
				else{
					return ("Debes pagar la consulta antes de pedir los medicamentos");
				}
			}
		}
		return domicilio;
			
	}
}

File: Python/src/gestorAplicacion/personas/Paciente.py

#Clase Paciente, para poder almacenar los atributos y aplicar los metodos respecto a los pacientes del sistema se usan arraylist para almacenar todos los pacientes creados y para tener las lista de los síntomas que tenga el paciente.
#Al igual se usan tree maps, donde se guardan las consultas con su respectiva fecha y le consulta asignada y las entregas, donde se guardan las fechas y las netregas correspondientes.

from gestorAplicacion.personas.Administrador import Administrador
from gestorAplicacion.personas.Persona import Persona
import datetime as dt


class Paciente(Persona):

#constructores	
    def __init__(self, nombre, apellido, documento, numeroDocumento, edad, genero, telefono, correoElectronico, 
        direccion, eps, historicaClinica, pagado,sintomas = None, consultas = None, entregas = None):
        super().__init__(nombre, apellido, documento, numeroDocumento, edad, genero, telefono, correoElectronico, direccion)
        if sintomas is None:
            sintomas = []
        if consultas is None:
            consultas = {}
        if entregas is None:
            entregas = {}
        self.eps = eps
        self.sintomas = sintomas
        self.historiaClinica = historicaClinica
        self.consultas = consultas 
        self.entregas = entregas
        self.pagado = pagado

	#Metodos
    #Getters y setters

    def getEps(self):
        return self.eps
    def setEps(self, eps):
        self.eps = eps

    def getSintomas(self):
        return self.sintomas

    def setSintomas(self, sintomas):
        self.sintomas = sintomas

    def getHistoriaClinica(self):
        return self.historiaClinica
    
    def setHistoriaClinica(self, historiaClinica):
        self.historiaClinica = historiaClinica

    def getConsultas(self):
        return self.consultas

    def setConsultas(self, consultas):
        self.consultas = consultas

    def isPagado(self):
        return self.pagado

    def setPagado(self, pagado):
        self.pagado = pagado
        
    def getEntregas(self):
        return self.entregas
	
    def setEntregas(self, entregas):
        self.entregas = entregas
        
    #Metodos
    '''
 	/*Metodo para poder ejecutar la funcionalidad de pedir cita, el cual recibe la fecha en que se desea pedir la cita, el tipo de la cita, el motivo y el típo de médico 
 	requerido, este retorna un String que indica si el proceso fue exitoso o fallido, en el primer caso, a su vez se devuelve la fecha en que fue asignada la cita, junto con
 	el consultorio en el que será la cita.*/
	'''
    def pedirCita(self, fecha, tipoCita, motivo, tipoMedico):
        medicos = Administrador.getMedicos()
        medico = None

        for i in medicos:
            #Se recorre la lista de los medicos creados y se va verificando si la especialidad del medico i concuerda con el tipo de cita que
 			#se requiere y tambien se verifica si ese medico esta disponible.
            if i.getEspecialista()== tipoMedico and Administrador.verificarDisponibilidadMedico(fecha, i) == True:
                medico = i
                break
        
        consultorios = Administrador.getConsultorios()
        consultorio = None
        for i in consultorios:
            #Se recorre la lista de los consultorios creados y se verifica si el consultorio esta disponible en la fecha que se requiere la cita
            if Administrador.verificarDisponibilidadConsultorio(fecha, i) == True:
                consultorio = i
                break

        if medico != None and consultorio != None:
            Administrador.asignarCita(self, medico, consultorio, fecha, motivo, tipoCita)
            return f"Su cita ha sido agendada exitosamente el: {fecha} con el médico: {medico.nombre} {medico.apellido} en el consultorio {consultorio.id} \n" 
        else:
            return "Hubo un error agendando su cita, vuelvalo a intentar. \n"

        '''
        Método para reagendar una cita, el cual recibe el identificador de la cita que se desea modificar y la nueva fecha a la que se desea hacer el cambio, el retornno de esta
        funcion es un String que muestra si el reagendamiento fue exitoso o no, el caso de exito hay 3 diferentes posibilidades, la primera es en caso que tanto el medico como el
        consultorio esten disponibles en la nueva fecha, el otro caso es que el consultorio esté disponible, pero no el medico, en este caso se retorna si se pudo reagendar, 
        pero hbuo cambio de medico y se muestra el nombre y apellido del nuevo medico y el caso final es que ni el consultorio ni el medico original se deben de cambiar,
        en este caso se retorna que el cambio de cita fue exitos, sin embargo que se cambio tanto el consultorio como el medico, mostrando los respectivos cambios*/
        '''

    def reagendarCita(self, id, fecha):
        cita = None
        citas = Administrador.getCitas()
        #Se busca la cita que se quiere reagendar
        for i in citas:
            if i.getPaciente() == self and i.getId() == id:
                cita = i
                break
        if cita is None:
            return "Hubo un problema reagendando su cita, vuelvalo a intentar más tarde"

        #Se verifica que tanto el medico como el consultorio esten disponible en la nueva fecha
        if Administrador.verificarDisponibilidadMedico(fecha, cita.getMedico()) == True and Administrador.verificarDisponibilidadConsultorio(fecha, cita.getConsultorio()) == True:
            cita.setFecha(fecha)
            return "Su cita ha sido reagendada exitosamente y será en el mismo consultorio, con el mismo médico"
            
        elif Administrador.verificarDisponibilidadMedico(fecha, cita.getMedico()) == False:
            #Se busca un medico con disponibilidad
            medicos = Administrador.getMedicos()
            medico = None
            for i in range(len(medicos)):
                if medicos[i].getEspecialidad() == cita.getTiposCitas() and Administrador.verificarDisponibilidadMedico(fecha, medicos[i]) == True:
                    medico = medicos[i]
                    cita.setMedico(medico)
                    cita.setFecha(fecha)
                    if Administrador.verificarDisponibilidadConsultorio(fecha, cita.getConsultorio) == True and medico != None:
                        return f"Su cita ha sido reagendada exitosamente, su nuevo medico será {medico.getNombre()} {medico.getApellido()} en el mismo consultorio \n"        
                    #Si el conusltorio no esta disponible, se busca un nuevo consultorio
                    else:
                        consultorios = Administrador.getConsultorios()
                        consultorio = None
                        for i in consultorios:
                            #Se recorre la lista de los consultorios creados y se verifica si el consultorio esta disponible en la fecha que se requiere la cita
                            if Administrador.verificarDisponibilidadConsultorio(fecha, i) == True:
                                consultorio = i
                                cita.setConsultorio(consultorio)
                                break
                        return f"Su cita ha sido reagendada exitosamente, su nuevo medico {medico.getNombre()} {medico.getApellido()} y esta será en el consultorio: {consultorio.getId()} \n"

                elif Administrador.verificarDisponibilidad(fecha, medicos[i]) == False and i == len(medicos)-1:
                    return "No fue posible reagendar su cita, debido a que no hay disponibilidad de citas, vuelvalo a intentar mas tarde"

        return "Hubo un problema reagendando su cita, vuelvalo a intentar mas tarde"            
    
    def solicitarExamen(examen, tipoMedico, medicos):
        if Administrador.autorizarExamen(examen, tipoMedico, medicos) == 1:
            return 1
        elif Administrador.autorizarExamen(examen, tipoMedico, medicos) == 2:
            return 2
        else:
            return 0
    
    